servers/quarkus-server/src/main/java/org/projectnessie/server/authz/CelAccessChecker.java

```diff
@@ -0,0 +1,186 @@
+  public void canReadEntityValue(AccessContext context, NamedRef ref, ContentsKey key)
+      throws ForbiddenException {
+    canPerformOpOnPath(context, key, AuthorizationRuleType.READ_ENTITY_VALUE);
+  }
+
+  @Override
+  public void canUpdateEntity(AccessContext context, NamedRef ref, ContentsKey key)
+      throws ForbiddenException {
+    canPerformOpOnPath(context, key, AuthorizationRuleType.UPDATE_ENTITY);
+  }
+
+  @Override
+  public void canDeleteEntity(AccessContext context, NamedRef ref, ContentsKey key)
+      throws ForbiddenException {
+    canPerformOpOnPath(context, key, AuthorizationRuleType.DELETE_ENTITY);
+  }
+
+  private void canPerformOpOnReference(
+      AccessContext context, NamedRef ref, AuthorizationRuleType type) {
+    if (!config.enabled()) {
+      return;
+    }
+    boolean allowed =
+        config.rules().entrySet().stream()
+            .anyMatch(
+                entry -> {
+                  try {
+                    return CELUtil.SCRIPT_HOST
+                        .buildScript(entry.getValue().replace("\"", ""))
+                        .withContainer(CELUtil.CONTAINER)
+                        .withDeclarations(CELUtil.AUTHORIZATION_RULE_DECLARATIONS)
+                        .build()
+                        .execute(
+                            Boolean.class,
+                            ImmutableMap.of(
+                                "ref",
+                                ref.getName(),
+                                "role",
+                                getRoleName(context),
+                                "op",
+                                type.name()));
+                  } catch (ScriptException e) {
+                    throw new ForbiddenException(
+                        String.format(
+                            "Failed to compile query expression with id '%s' and expression '%s' due to: %s",
+                            entry.getKey(), entry.getValue(), e.getMessage()));
+                  }
+                });
+    if (!allowed) {
+      throw new ForbiddenException(
+          String.format(
+              "'%s' is not allowed for Role '%s' on Reference '%s'",
+              type, getRoleName(context), ref.getName()));
+    }
+  }
+
+  private String getRoleName(AccessContext context) {
+    return null != context.user() ? context.user().getName() : "";
+  }
+
+  private void canPerformOpOnPath(
+      AccessContext context, ContentsKey contentsKey, AuthorizationRuleType type) {
+    if (!config.enabled()) {
+      return;
+    }
+    boolean allowed =
+        config.rules().entrySet().stream()
+            .anyMatch(
+                entry -> {
+                  try {
+                    return CELUtil.SCRIPT_HOST
+                        .buildScript(entry.getValue().replace("\"", ""))
+                        .withContainer(CELUtil.CONTAINER)
+                        .withDeclarations(CELUtil.AUTHORIZATION_RULE_DECLARATIONS)
+                        .build()
```

Comment:
The built scripts should really be held in a "global" map (like a ConcurrentHashMap using computeIfAbsent), because parsing and checking the script every time an access needs to be checked it too expensive.

Might be worth a PR against cel-java, which could then use for example caffeine instead of a "stupid CHM" ;)

Reply:
If (really, *if*) the CEL expressions are rather constant (like defined in the configuration at Quarkus-image-build-time), the actual CEL parse+check operations should ideally be part of the Quarkus-image-build and not slow down the startup.

Reply:
I didn't focus too much on perf in this iteration, but I totally agree that this should be cached as there's no need to parse/check scripts all the time, even though they don't change during runtime

Reply:
added `CompiledAuthorizationRules` that compiles the rules during startup